### PR TITLE
Unify PhpCompilerAttribute + PhpAttribute; allow attributes on internal classes + fns

### PR DESCRIPTION
**Pull request by beberlei**
@kooldev Simplify and adds attributes on everything.

- [x] Removed `PhpCompilerAttribute`, each "compiler" attribute is just regular now.
- [x] Add attributes on internal classes
- [x] Add attributes on internal functions/methods
- [x] Add test showing that fetching attribues on `PhpAttribute` returns itself a `PhpAttribute`. Still failing, because it doesn't detect this properly on newInstance yet.
- [x] Add a compiler attribute example to zend_test

### CODE REVIEW

File: Zend/zend_attributes.c [hunk 1]
```diff
@@ -70,12 +64,38 @@ ZEND_API zend_attribute *zend_get_attribute_str(HashTable *attributes, const cha
...
+static void attribute_ptr_dtor(zval *v) /* {{{ */
+{
+	zend_attribute_free((zend_attribute *) Z_PTR_P(v));
+}
+/* }}} */
+
 ZEND_API void zend_compiler_attribute_register(zend_class_entry *ce, zend_attributes_internal_validator validator)
 {
+	if (ce->type != ZEND_INTERNAL_CLASS) {
+		return;
```
Comment (kooldev) [inline Zend/zend_attributes.c]: This should throw an error using `zend_error_noreturn()` instead. This case will likely never happen, but it is a fatal error if it does.

File: Zend/zend_attributes.c [hunk 2]
```diff
@@ -70,12 +64,38 @@ ZEND_API zend_attribute *zend_get_attribute_str(HashTable *attributes, const cha
...
 
 	zend_hash_update_ptr(&internal_validators, lcname, validator);
 	zend_string_release(lcname);
+
+	if (ce->attributes == NULL) {
+		ce->attributes = pemalloc(sizeof(HashTable), 1);
+		zend_hash_init(ce->attributes, 8, NULL, attribute_ptr_dtor, 1);
+	}
+
+	zend_attribute *attr = emalloc(ZEND_ATTRIBUTE_SIZE(0));
```
Comment (kooldev) [inline Zend/zend_attributes.c]: This has to use `pemalloc(ZEND_ATTRIBUTE_SIZE(0), 1)` instead.

File: Zend/zend_attributes.c [hunk 3]
```diff
@@ -70,12 +64,38 @@ ZEND_API zend_attribute *zend_get_attribute_str(HashTable *attributes, const cha
...
+
+	attr->name = zend_string_copy(zend_ce_php_attribute->name);
+	attr->lcname = zend_string_tolower(attr->name);
```
Comment (kooldev) [inline Zend/zend_attributes.c]: This needs persistent allocation as well: `zend_string_tolower_ex(attr->name, 1)`

File: ext/reflection/php_reflection.c [hunk 4]
```diff
@@ -1110,6 +1110,10 @@ static int read_attributes(zval *ret, HashTable *attributes, zend_class_entry *s
 		return SUCCESS;
 	}
 
+	if (!attributes) {
```
Comment (kooldev) [inline ext/reflection/php_reflection.c]: This check is already performed in `reflect_attributes()` and the return is unreachable. It should be replaced with `ZEND_ASSERT(attributes != NULL)` and moved to the top instead.

File: ext/reflection/php_reflection.c [hunk 5]
```diff
@@ -6582,12 +6584,9 @@ ZEND_METHOD(ReflectionAttribute, newInstance)
 		RETURN_THROWS();
 	}
 
-	if (ce->type == ZEND_USER_CLASS && !zend_get_attribute_str(ce->info.user.attributes, ZEND_STRL("phpattribute"), 0)) {
+	if (!zend_get_attribute_str(ce->attributes, ZEND_STRL("phpattribute"), 0)) {
```
Comment (kooldev) [inline ext/reflection/php_reflection.c]: The `offset` parameter (0) has to be dropped after #13 is merged.

**Review by kooldev — COMMENTED** [review]
Looking good so far. I added some comments mostly related to use of persistent memeory allocation. Letting go of compiler attribute is a good thing and supporting it on all internal classes / functions is also pretty nice.

File: ext/reflection/php_reflection.c [hunk 6]
```diff
@@ -4181,10 +4185,8 @@ ZEND_METHOD(ReflectionClass, getAttributes)
 
 	GET_REFLECTION_OBJECT_PTR(ce);
 
-	if (ce->type == ZEND_USER_CLASS && ce->info.user.attributes) {
-		attributes = ce->info.user.attributes;
-		scope = ce;
-	}
+	attributes = ce->attributes;
+	scope = ce;
```
Comment (kooldev) [inline ext/reflection/php_reflection.c]: I think the additional aliasing here is not needes, just pass `ce->attributes` and `ce` to `reflect_attributes()` instead.

**Comment by nikic** [comment]
You might want to add some additional tests in the zend_test extension.

**Pull request merged**